### ISSUE TITLE
Complete design for trigram padding

### ISSUE THREAD

**Issue by BobWall23** [IronCoreLabs/search-helpers#4]
Need to write up the design for padding the number of trigrams produced by indexing a string so that it is harder to look at the search index and identify indexed strings.

**Comment by BobWall23** [comment]
Got the proposal for padding done - in
https://docs.google.com/document/d/1u0d9FtrmkH2hW7xekz_9kAaMHOfsAQYx6vhPjc926hc

**Comment by BobWall23** [comment]
One potential weakness of this scheme is that an attacker can retrieve some basic information without reversing the hash; one possible attack is doing frequency analysis on the index tokens. This allows the attacker to guess what certain tokens must represent due to their frequency of occurrence. Another, simpler attack is to look at the number of index terms that are produced for an entry - this provides a significant amount of information about the length of the string that was indexed. It is not exact, since the same trigram might occur multiple times in the string, but it is still easy to correlate that with possible strings that were indexed.

To mitigate, we will add padding terms when we index a string. These are just random 32-bit values that we add into the list of index terms for a string. There is a possibility that we choose a padding term that actually collides with a term that is generated from another string, but there is a possibility of collisions anyway, due to bucketing, so post-filtering should allow any false matches to be discarded by the client.

There are a number of different schemes we could use to choose the number of padding terms. One is to choose a maximum limit that we allow for index terms, and pad every string’s list of index terms to that length. As long as we choose a large enough value for the number of terms, this would absolutely prevent someone from extracting any information about the length of the indexed string by looking at the list of index terms. However, this would generate a large number of padding terms, which could in turn increase the number of false positives in search results.

Another approach is to select a random number from some probability distribution (uniform, normal, or some other distribution) and add that many padding terms. Analysis of the results may be able to determine the parameters of the distribution that is sampled (since the number of index terms for each string is the sum of samples from two different distributions), but it should obscure the length of any individual set of index terms (and thus the length of the indexed string).

A similar technique is to select a random number from a probability distribution and make the total number of index terms with padding match that value. A possible candidate is the log normal distribution. This distribution does not produce negative values and can be tuned to scatter results closer to or further from a mean value.

If we sample from this distribution, multiply the result by a constant and round, then add a constant minimum value, we can produce a target length for the padded index term list for a particular string. If this value is less than the number of index terms that were generated, we just resample and repeat until we get a value at least that large. This will cause the distribution of the generated lists of index terms to deviate slightly from the log normal distribution, but this isn’t significant.

Initially, we will generate values using this formula
    `10 + int(round(random.lognormvariate(0,.7) * 3))`

where `lognormvariate()` is a function that samples from the log normal distribution with parameter mu = 0 and sigma = 0.7.

**Comment by BobWall23** [comment]
Simplification that is much less sensitive to picking parameters for a distribution (and sampling a distribution other than uniform):

Given a list of trigrams that is no more than 199 long,

- 1% of the time, add a random padding of 1 - 199 trigrams.
- 3% of the time, add between 1 and 50 trigrams
- 21% of the time, add between 1 and 20 trigrams
- 75% of the time, add between 1 and 10 trigrams.

Cap the length of the padded trigram list at 200.

```
  prob = random.randint(1,100)
  if prob < 2:
    nextval = random.randint(1, 200)
  elif prob < 5:
    nextval = random.randint(1, 50)
  elif prob < 25:
    nextval = random.randint(1, 20)
  else:
    nextval = random.randint(1, 10)
  nextval = min((vals[i] + nextval), 200)
```

This introduces sufficient randomness to bump some of the shorter inputs up to long lists, helping to obscure any long outliers in the input, without generating overly large amounts of padding trigrams.

**Comment by BobWall23** [comment]
For an input of 20000 trigram lists, with the following distribution of lengths
```
Histogram:
[0, 0, 0, 174, 3859, 5005, 3755, 2401, 1565, 1032, 654, 492, 328, 193, 138, 113, 76, 65, 35, 25, 16, 21, 8, 8, 6, 3, 3, 5, 3, 3, 2, 1, 1, 0, 3, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0]
```
It produced this distribution of output list lengths:
```
[0, 0, 0, 0, 16, 337, 820, 1058, 1249, 1520, 1583, 1609, 1647, 1694, 1667, 1398, 1030, 724, 518, 452, 398, 322, 294, 237, 246, 219, 116, 105, 74, 62, 40, 29, 33, 14, 23, 21, 13, 18, 20, 12, 23, 16, 11, 21, 18, 13, 18, 13, 13, 4, 10, 8, 9, 10, 9, 14, 5, 7, 4, 5, 1, 1, 1, 1, 1, 4, 3, 0, 0, 1, 0, 0, 0, 1, 2, 0, 1, 2, 0, 1, 2, 2, 3, 0, 1, 0, 1, 0, 1, 0, 0, 2, 1, 1, 2, 0, 2, 1, 0, 1, 1, 0, 0, 3, 3, 2, 2, 2, 3, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 2, 0, 0, 2, 2, 0, 1, 0, 1, 3, 0, 2, 2, 2, 1, 1, 2, 3, 2, 0, 1, 1, 1, 0, 0, 0, 0, 0, 2, 0, 1, 1, 0, 0, 2, 0, 0, 2, 0, 5, 1, 1, 1, 0, 2, 0, 2, 0, 0, 3, 0, 0, 2, 1, 0, 0, 0, 4, 1, 2, 4, 1, 1, 1, 0, 1, 0, 0, 2, 2, 1, 0, 0, 2, 2, 2, 0, 1, 2, 1, 5]
```

**Comment by BobWall23** [comment]
For another run, the total number of input trigrams was 130141
The total number of output trigrams was 291860 - a 124% increase

**Comment by BobWall23** [comment]
Tweaking the percentages and padding,
```
  prob = random.randint(1,200)
  if prob < 2:
    padlen = random.randint(1, 199)
  elif prob < 6:
    padlen = random.randint(1, 30)
  elif prob < 51:
    padlen = random.randint(1, 10)
  else:
    padlen = random.randint(1, 5)
  padlen = min(200 - inputlen, padlen)
```
generates
```
Input length histogram:
[0, 0, 0, 167, 3844, 5243, 3709, 2436, 1527, 988, 667, 418, 262, 232, 140, 93,
 48, 47, 44, 33, 22, 21, 8, 8, 8, 8, 4, 4, 5, 4, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0,
 2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0]
Output length histogram:
[0, 0, 0, 0, 23, 665, 1642, 2172, 2687, 2902, 2620, 1836, 1330, 1027, 858,
 581, 428, 260, 185, 147, 112, 68, 62, 46, 32, 32, 23, 32, 26, 12, 21,
 22, 18, 18, 13, 7, 5, 8, 7, 3, 2, 2, 1, 0, 1, 3, 0, 1, 0, 0, 2, 0, 1,
 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0,
 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 2, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0,
 0, 0, 1, 0, 0, 1, 1, 0, 0, 2, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1,
 0, 2, 1, 0, 0, 2, 0, 0, 2, 1, 0, 0, 0, 0, 0, 0, 1, 3, 0, 0, 1, 0,
 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0,
 0, 2, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 4]
Total input trigrams: 128892
Total output trigrams: 212328
```
a 65% increase, but it still looks usable.

**Comment by coltfred** [comment]
Implemented as #6
